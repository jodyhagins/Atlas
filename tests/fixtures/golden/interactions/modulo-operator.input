guard_prefix=TEST
guard_separator=_
upcase_guard=true

# Modulo is integer-only and has different semantics than other arithmetic
namespace=data
constexpr
value_access=.value

ByteCount % ByteCount -> ByteCount
ByteCount % size_t -> ByteCount
