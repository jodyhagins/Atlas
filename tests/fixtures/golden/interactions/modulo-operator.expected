#ifndef TEST_279301F87A200A876D26003DEA3C8AFBCA1F439F
#define TEST_279301F87A200A876D26003DEA3C8AFBCA1F439F

// ======================================================================
// NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE
// ----------------------------------------------------------------------
//
// DO NOT EDIT THIS FILE DIRECTLY.
//
// This source file has been generated by Atlas Interaction Generator v0.1.0
// https://github.com/jodyhagins/Atlas
//
// DO NOT EDIT THIS FILE DIRECTLY.
//
// ----------------------------------------------------------------------
// NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE
// ======================================================================

#if __has_include(<version>)
#include <version>
#endif

#include <type_traits>
#include <utility>

#ifndef WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90
#define WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90

// ======================================================================
// ATLAS STRONG TYPE BOILERPLATE
// ----------------------------------------------------------------------
//
// This section provides the infrastructure for Atlas strong types.
// It is identical across all Atlas-generated files and uses a shared
// header guard (WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90) to ensure
// the boilerplate is only included once even when multiple generated
// files are used in the same translation unit.
//
// The boilerplate is intentionally inlined to make generated code
// self-contained with zero external dependencies.
//
// Components:
// - atlas::strong_type_tag: Base class for strong types
// - atlas::value(): Universal value accessor for strong types
// - atlas_detail::*: Internal implementation utilities
//
// For projects using multiple Atlas-generated files, this boilerplate
// will only be compiled once per translation unit thanks to the shared
// header guard below.
//
// ----------------------------------------------------------------------
// DO NOT EDIT THIS SECTION
// ======================================================================

// Atlas feature detection macros
#ifndef ATLAS_NODISCARD
#if defined(__cpp_attributes) && __cpp_attributes >= 201603L
#define ATLAS_NODISCARD [[nodiscard]]
#else
#define ATLAS_NODISCARD
#endif
#endif

#if defined(__cpp_impl_three_way_comparison) && \
    __cpp_impl_three_way_comparison >= 201907L
#include <compare>
#endif

namespace atlas {

struct strong_type_tag
{
#if defined(__cpp_impl_three_way_comparison) && \
    __cpp_impl_three_way_comparison >= 201907L
    friend auto operator <=> (
        strong_type_tag const &,
        strong_type_tag const &) = default;
#endif
};

struct value_tag
{ };

namespace atlas_detail {

template <typename... Ts>
struct make_void
{
    using type = void;
};

template <typename... Ts>
using void_t = typename make_void<Ts...>::type;

template <typename T, typename = void>
struct IsAtlasType
: std::false_type
{ };

template <typename T>
struct IsAtlasType<T, void_t<typename T::atlas_value_type>>
: std::true_type
{ };

template <std::size_t N>
struct PriorityTag
: PriorityTag<N - 1>
{ };

template <>
struct PriorityTag<0u>
{ };

using value_tag = PriorityTag<3>;

template <bool B>
using bool_c = std::integral_constant<bool, B>;
template <typename T>
using bool_ = bool_c<T::value>;
template <typename T>
using not_ = bool_c<not T::value>;
template <typename T, typename U>
using and_ = bool_c<T::value && U::value>;
template <typename T>
using is_lref = std::is_lvalue_reference<T>;
template <typename T, typename U = void>
using enable_if = typename std::enable_if<T::value, U>::type;

template <typename T>
using _t = typename T::type;

void atlas_value();

template <typename T>
constexpr T &
value(T & val, PriorityTag<0>)
{
    return val;
}

template <typename T, typename U = typename T::atlas_value_type>
using val_t = _t<std::conditional<std::is_const<T>::value, U const &, U &>>;

template <typename T, typename U = val_t<T>>
constexpr auto
value(T & val, PriorityTag<1>)
-> decltype(atlas::atlas_detail::value(static_cast<U>(val), value_tag{}))
{
    return atlas::atlas_detail::value(static_cast<U>(val), value_tag{});
}

template <typename T>
constexpr auto
value(T const & t, PriorityTag<2>)
-> decltype(atlas_value(t, atlas::value_tag{}))
{
    return atlas_value(t, atlas::value_tag{});
}

template <typename T>
constexpr auto
value(T const & t, PriorityTag<3>)
-> decltype(atlas_value(t))
{
    return atlas_value(t);
}

class Value
{
    template <
        typename U,
        typename T,
        typename V = _t<std::conditional<is_lref<U &&>::value, T &, T>>>
    static constexpr V rval(T && t)
    {
        return t;
    }

public:
    template <typename T>
    constexpr auto operator () (T && t) const
    -> decltype(rval<T>(atlas_detail::value(t, atlas_detail::value_tag{})))
    {
        return rval<T>(atlas_detail::value(t, atlas_detail::value_tag{}));
    }
};

void begin();
void end();

template <typename T>
constexpr auto
begin_(T && t) noexcept(noexcept(begin(std::forward<T>(t))))
-> decltype(begin(std::forward<T>(t)))
{
    return begin(std::forward<T>(t));
}

template <typename T>
constexpr auto
end_(T && t) noexcept(noexcept(end(std::forward<T>(t))))
-> decltype(end(std::forward<T>(t)))
{
    return end(std::forward<T>(t));
}

} // namespace atlas_detail

#if defined(__cpp_inline_variables) && __cpp_inline_variables >= 201606L
inline constexpr auto value = atlas_detail::Value{};
#else
template <typename T>
constexpr auto
value(T && t)
-> decltype(atlas_detail::Value{}(std::forward<T>(t)))
{
    return atlas_detail::Value{}(std::forward<T>(t));
}
#endif

} // namespace atlas

#endif // WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90


//////////////////////////////////////////////////////////////////////
///
/// These are the droids you are looking for!
///
//////////////////////////////////////////////////////////////////////


// Custom value accessors for non-Atlas types
// These allow atlas::value() to work with external library types
// Users can override by providing atlas_value(T const&) without the tag parameter
namespace atlas {
inline constexpr auto
atlas_value(::data::ByteCount const& v, value_tag)
-> decltype(v.value)
{
    return v.value;
}

} // namespace atlas


// Compound assignment operators for cross-type interactions
// These use ADL to be found automatically for atlas strong types
// The decltype ensures they only match when the binary operator is defined
namespace atlas {

namespace atlas_detail {
template <typename L, typename R, typename = void>
struct has_compound_op_modulo
: std::false_type
{ };

template <typename L, typename R>
struct has_compound_op_modulo<
    L,
    R,
    decltype((void)(atlas::value(std::declval<L&>()) %=
        atlas::value(std::declval<R const&>())))>
: std::true_type
{ };

template <typename L, typename R>
constexpr L &
compound_assign_impl_modulo(L & lhs, R const & rhs, std::true_type)
noexcept(noexcept(atlas::value(lhs) %= atlas::value(rhs)))
{
    atlas::value(lhs) %= atlas::value(rhs);
    return lhs;
}

template <typename L, typename R>
constexpr L &
compound_assign_impl_modulo(L & lhs, R const & rhs, std::false_type)
noexcept(noexcept(atlas::value(lhs) = atlas::value(lhs % rhs)))
{
    atlas::value(lhs) = atlas::value(lhs % rhs);
    return lhs;
}
}

template <
    typename L,
    typename R,
    typename std::enable_if<
        std::is_base_of<atlas::strong_type_tag, L>::value,
        bool>::type = true>
inline auto
operator%=(L & lhs, R const & rhs)
noexcept(noexcept(atlas_detail::compound_assign_impl_modulo(
    lhs,
    rhs,
    atlas_detail::has_compound_op_modulo<L, R>{})))
-> decltype(atlas_detail::compound_assign_impl_modulo(
    lhs,
    rhs,
    atlas_detail::has_compound_op_modulo<L, R>{}))
{
    return atlas_detail::compound_assign_impl_modulo(
        lhs,
        rhs,
        atlas_detail::has_compound_op_modulo<L, R>{});
}

} // namespace atlas


//////////////////////////////////////////////////////////////////////
///
/// These are the droids you are looking for!
///
//////////////////////////////////////////////////////////////////////
namespace data {

inline constexpr ByteCount
operator%(ByteCount lhs, ByteCount rhs)
noexcept(
    noexcept(lhs.value % rhs.value) &&
    std::is_nothrow_constructible<ByteCount, decltype(lhs.value % rhs.value)>::value)
{
    return ByteCount{lhs.value % rhs.value};
}

inline constexpr ByteCount
operator%(ByteCount lhs, size_t rhs)
noexcept(
    noexcept(lhs.value % rhs) &&
    std::is_nothrow_constructible<ByteCount, decltype(lhs.value % rhs)>::value)
{
    return ByteCount{lhs.value % rhs};
}

} // namespace data

#endif // TEST_279301F87A200A876D26003DEA3C8AFBCA1F439F

